### ISSUE TITLE
I have *the* weirdest issue I've ever seen with Linux, ZFS and SAMBA

### ISSUE THREAD

**Issue by Rudd-O** [openzfs/zfs#14391]
Forgive me for posting this, it's a complicated bug and I'm at the end of my wits (5 hours and no progress here).

I am using Fedora 36 with SAMBA samba-4.16.8-1.fc36.x86_64 , ZFS c935fe2e9267b2ea6e3187e205aac019f5e0c83f and kernel 6.0.16-200.fc36.x86_64 .

Today I moved a file from a Windows 10 client (all up-to-date) to its desktop, and then I realized I had to copy the exact file back to the server.  This is on a share that's backed by a ZFS file system.

Surprisingly, Windows told me that the server had no space (I'm paraphrasing — it's in German) and it asked me if I wanted to replace the file.  Wut, there is not supposed to be any file there.  I said to replace, and I got an error — could not create the file because "the file is too large".  This wasn't a big file (couple hundred kilobytes), so this surprised me.  Also, on the server, a zero-byte file with *exactly* the same file name was created.

I tried copying another file from the desktop, and that caused no problem.  I then copied back that file from the server to the desktop, and finally copied the file again from the desktop back to the server — boom!  Zero-byte file, same "too large" error message.  So it looks like there's something "sticky" on the files I copy from the file server, such that when I copy them back to the file server, I get these problems.

I tried with `xcopy` too.  I get `File creation error -- parameter incorrect`.  On the SAMBA server side, the log shows the client trying to open the file 8 times, and then says this:

```
  smbd_smb2_request_error_ex: smbd_smb2_request_error_ex: idx[1] status[NT_STATUS_INVALID_PARAMETER] || at ../../source3/smbd/smb2_create.c:338
```

which is consistent with the error I am getting.

How weird is this?  Remember, **only files whose origin was the file server have this problem when copying them back to the file server**.  I've never seen anything like that, and the files that fail don't have any weird ACLs (they are the same as the ACLs of the files that copy successfully) nor do they have any NTFS streams either.

Furthermore, this issue is only reproducible on Windows 10 clients.  Android and Linux clients have no such issue.

Following a hunch, I created a file share to a `tmpfs` file system and attempted the copy again.  This time, success.  Then I created another file share to an `ext4` file system on the same machine — total success in copying the files.  This indicates to me that there's a problem specifically with ZFS and SAMBA that causes Windows 10 to get "invalid parameter" when creating certain files — _but only specifically certain files which came from the file server to begin with_.

What the heck!  What could it be?  How could I debug this harder?  Thanks in advance.

This is the file server configuration in question.  I have tweaked settings but no difference.

```
[global]

log level = 3
debug class = yes
debug uid = yes
max log size = 0
workgroup = DRAGONFEAR
server string = DragonFear file server
log file = /var/log/samba/log.%U.%m.log

map to guest = never

printing = lprng
printcap name = /etc/printcap
load printers = no

security = user
passdb backend = tdbsam

load printers = no

# printer sucks
server min protocol = SMB3_11

cups options = raw

guest account = guest
map archive = no
map system = no
map hidden = no

[homes]
comment = Home directories
browseable = no
valid users = %S
writable = yes
guest ok = no
public = no
vfs objects = catia
catia:mappings = 0x22:0xa8,0x2a:0xa4,0x2f:0xf8,0x3a:0xf7,0x3c:0xab,0x3e:0xbb,0x3f:0xbf,0x5c:0xff,0x7c:0xa6
nt acl support = yes

```

**Comment by Rudd-O** [comment]
Seriously you have to see this.  Hahaha!  WTF!

![PXL_20230113_172413370_1](https://user-images.githubusercontent.com/346306/212381322-0cc58951-aa28-431b-96a4-b0780c375fa9.jpg)

**Comment by Rudd-O** [comment]
The SAMBA code path in question: https://github.com/samba-team/samba/blob/master/source3/smbd/smb2_create.c#L338

**Comment by rincebrain** [comment]
To debug this further, I'd probably suggest tracing the `open()` call samba is making, and then all calls associated with that fd to things like write/fcntl/whatever, and seeing which go boom.

Alternately, you could start instrumenting the ZFS code for file creation and read/write/the like, with debug prints or tracing, and see where it returns an error...

My best guess, if it's something to do with files that originated on that ZFS dataset, is that it's something with the ACLs being passed around being unreproducible somehow. Either that or something in 6.0 changed and broke an API in some way that only Samba cares about.

**Comment by justinpryzby** [comment]
Is it an issue if you use a different destination filename than the on that was moved/removed ?

Also, should indicate if zfs is using deduplication/compression, and show its "zpool list".

**Comment by IvanVolosyuk** [comment]
I would look at the acls of the source files on the windows machine. I'm pretty sure the files copied from the linux machine have different acls than those for files created on the windows machine. Try changing them?

Also try to play with removing one or more of the parameters:
```
vfs objects = catia
catia:mappings = 0x22:0xa8,0x2a:0xa4,0x2f:0xf8,0x3a:0xf7,0x3c:0xab,0x3e:0xbb,0x3f:0xbf,0x5c:0xff,0x7c:0xa6
nt acl support = yes
```
What are the values zfs properties you have for the dataset:
utf8only, normalization, casesensitivity, acltype, aclmode, aclinherit?

**Comment by Rudd-O** [comment]
> Is it an issue if you use a different destination filename than the on that was moved/removed ?
> 
> Also, should indicate if zfs is using deduplication/compression, and show its "zpool list".

Yes.  I tried renaming the file locally before uploading.  No solution.

> I would look at the acls of the source files on the windows machine. I'm pretty sure the files copied from the linux machine have different acls than those for files created on the windows machine. Try changing them?

ACLs were identical when comparing the failing files and files originating from the Windows machine that nonetheless successfully copied.

> What are the values zfs properties you have for the dataset: utf8only, normalization, casesensitivity, acltype, aclmode, aclinherit?

I don't think that matters — the same Windows machine has no problem copying other files that were *not* downloaded from the file server.

> Also try to play with removing one or more of the parameters:

I did.  No difference.  That was actually the first thing I tried.

**Comment by rincebrain** [comment]
I mean, the logical thing to do is to ask what makes the files distinct, I suppose.

If you copy the file to a new filename (not rename) then copy it back, does it break?
If you reboot and then copy it back, does it break?
If you copy it to a new filename from a Linux boot environment and ntfs-3g, and then copy that back, does it break?

And so on.

My other guess if it's not ACLs is that it's trying to preserve ADS and failing.

**Comment by IvanVolosyuk** [comment]
NTFS also have named streams for the same filename, which often used by
antiviruses to indicate if the origin is trustworthy, etc.

**Comment by rincebrain** [comment]
ADS, or alternate data streams, are that feature, yes.

**Comment by Rudd-O** [comment]
I looked at streams with the `dir` command line switch.  No streams.  That's why I'm so baffled.  What else could it be?

> I mean, the logical thing to do is to ask what makes the files distinct, I suppose.

100% in agreement.

> If you copy the file to a new filename (not rename) then copy it back, does it break?
> If you reboot and then copy it back, does it break?
> If you copy it to a new filename from a Linux boot environment and ntfs-3g, and then copy that back, does it break?
>
> And so on.

I think I will try these very soon (this week).

> My other guess if it's not ACLs is that it's trying to preserve ADS and failing. (ADS is how Windows pronounces what, if you used old macOS, you might call Resource Forks there - and IIRC ZFS would shove them into xattrs, but xattrs on Linux have a very finite limit...)

No ADS on the files.

**Comment by devZer0** [comment]
did you check windows event log for anything noticeable ?

**Comment by Rudd-O** [comment]
Sadly no.  Will have to retest sometime this weekend hopefully.

**Comment by aasraoui** [comment]
Has this been resolved or should we reopen it ?

thanks!

**Comment by Rudd-O** [comment]
Regrettably I don't have the bandwidth to test.  Please feel free to close.  I will reopen in the future if I encounter the issue again.
